### ISSUE TITLE
Avatar became inaccessible after the domain was changed.

### ISSUE THREAD

**Issue by Octobug** [apache/incubator-answer#861]
## Describe the bug

Avatar became inaccessible after the domain was changed. I only tested it in development env.

### To Reproduce

1. Under `domain.old`, you uploaded your custom avatar.
2. You updated `domain.old` to `domain.new`, and the uploaded avatar became inaccessible.
    - Domain changes are very common.

### Expected behavior

Uploaded files should serve well.

### Screenshots

<img width="829" alt="image" src="https://github.com/apache/incubator-answer/assets/8007022/770d64b1-095f-4c0c-b61d-854d40ce1247">

<img width="679" alt="image" src="https://github.com/apache/incubator-answer/assets/8007022/520afe2f-c098-4fa9-a255-ca5e153567d3">

### Platform

- Device: Desktop
- OS: macOS Ventura 13.6.1
- Browser and version: Chrome 122.0.6261.129 (Official Build) (x86_64)
- Version: v1.2.5

**Comment by Octobug** [comment]
> Thanks for the feedback. It seems that the storing avatars relies on absolute paths. Modifying it to a new domain could potentially introduce new issues. Saving them as relative paths might be a better option. 🤔

@LinkinStars 
Yeah, uploaded files should be stored as relative paths. And, for those already deployed in production, the fixing should implement a mechanism (maybe an independent command tool) to automatically update existing data.

**Comment by Octobug** [comment]
@LinkinStars I found that all images in a question, answer or avatar have this issue. I've changed the issue title.

**Comment by Octobug** [comment]
I would like to solve this issue once the solution is settled. 👀

**Comment by calmdev** [comment]
Fwiw, I ran into this issue with logo and favicon as well under the branding settings when originally testing how to go about setting up answer.

**Comment by Octobug** [comment]
@LinkinStars Thank you for your detailed response!

⚠️ I haven't read much of the relevant code yet so I might miss some background knowledge of Answer. I'll make a summary to see if I have understood correctly.

I am also in favor of building this functionality as a `./answer` command option.

And let's split a URL into 2 major parts for the convenience in describing:

- `PROTOCOL://HOSTNAME:PORT`
- `PATH`
  - `BASE_PATH`
  - `RELATIVE_PATH`

I think the `ABSOLUTE_IMAGE_PATH` could be renamed to `IMAGE_PATH_PREFIX` (if there will be other types of files in the future, maybe it could be `UPLOADS_PATH_PREFIX`).

There are 2 major types of uploaded files:

1. For any external storage provided by 3rd-party services (S3, OSS, ...), files are stored as `PROTOCOL://HOSTNAME:PORT + PATH` in DB with no doubt.
2. For files stored locally, application codes can directly use `IMAGE_PATH_PREFIX + RELATIVE_PATH`, only `RELATIVE_PATH` is stored in DB. This usage of `IMAGE_PATH_PREFIX` can be delivered by documentation:
    - 2.1. By default, the value of `IMAGE_PATH_PREFIX` is `/`
    - 2.2. For files need to be moved to a new directory, the value of `IMAGE_PATH_PREFIX` is `BASE_PATH` (e.g. `new/dirname/`)
    - 2.3. For files need to be moved to another self-hosted place different from the Answer instance, the value of `IMAGE_PATH_PREFIX` is `PROTOCOL://HOSTNAME:PORT + BASE_PATH` (e.g. `https://statics.self-hosted.com/uploads/`)

Pseudocode:

```js
// When fetching existing data to rewrite:
getImageURL(image) {
    if (isSelfHosted(image)) {
        return IMAGE_PATH_PREFIX + image.path;
    } else {
        return image.path;
    }
}
```

As for the `./answer fix image from a to b` command, I think the `from a` and `to b` args are not needed. Because what are stored in DB are always `RELATIVE_PATH`s.

- Or, are the two args used for auto-migrating uploaded files (`mv SRC DST`)?
- The description above assumes that an admin move uploaded files manually.

**Comment by Octobug** [comment]
This task seems heavy for a Go & Answer beginner (exactly me 🤣).

If the coding job is real complicated under your estimation, I think it's better be done by your team.

**Comment by Octobug** [comment]
> Take it easy. You can try your best, and we will help you solve any problems together. 💪

@LinkinStars Okay then, I'll give it another try.

**Comment by Octobug** [comment]
@LinkinStars 

I have another thought on **external storage type**...

If the uploads dataset is huge, the migration could be very time-consuming because it needs to query & filter out **all** urls of avatars and urls in texts of questions & answers, and then write them back to DB.

What if Answer saves all types of uploads with relative paths with a prefix marker? Like: `:NUM:/uploads/post/57pxxvykQFj.jpeg`

`:NUM:` is a storage url prefix version number. And the version number maps to a URL prefix string, which is used to replace the `:NUM:` marker in a relative path stored in DB.

The default value of `NUM` is `0` mapping to `"/"`, it deals with local uploads type. When an Answer admin decides to move the `uploads` directory to another path, the admin just needs to change `0` to map to `"/new/path"` after files are settled down.

If the admin adds an external storage, since there is a `visit_url_prefix` config field for every storage plugin, Answer knows if it is a new prefix. If it is, it increases `NUM`, and set it as active so that any subsequent uploads use the new active `NUM`.

Let's say there is a new type of `site_info` called `uploads`, it looks like:

```json
{
    "url_prefix_versions": {
        // ":0:/uploads/post/57pxxvykQFj.jpeg"
        "0": "/",

        // ":1:/uploads/post/57pxxvykQFj.jpeg"
        "1": "https://s3.east-1.xxx.com/",

        // ":2:/uploads/post/57pxxvykQFj.jpeg"
        "2": "https://abc.oss-cn-hangzhou.aliyuncs.com/",

        ...
    },

    "url_prefix_versions_active": 1,
    // or [1, 2, ...] if Answer is to support multiple storages at the same time

    // other properties...
}
```

With this solution, those relative paths hardly need to be changed.

The basic migrate operation process is:

1. The admin copies their uploads to another place.
2. Then the admin updates the corresponding `NUM`'s mapping.
3. The admin now can delete uploads in the old place.

And still, `./answer migrate image from_a to_b` could be treated as the last resort to rewrite paths in DB. But if the version number solution is adopted, this migrate functionality is better be integrated in the admin UI. It can provide migration status like progress, affected number of urls, etc.

However, migration of large dataset may not be common. The consideration is whether it is worth implementing such a feature to address a rare scenario.

If this idea is rejected, I will try to implement the previously discussed solution.

**Comment by Octobug** [comment]
@LinkinStars Okay, it was just a thought. 👀 

I would like to confirm some details:

#### The env var

Does this var need to be supported both in [Config File](https://answer.apache.org/docs/configfile) and [Environment Variable](https://answer.apache.org/docs/env)?

- In config file: `service_config.upload_url_prefix`
  - Used when site is running.
- Environment Variable: `UPLOAD_URL_PREFIX`
  - Used for installation
- Default value: `"/"`

It actually affects all kinds of uploads, so I think `upload` is more proper. And it may include `PROTOCOL://HOSTNAME:PORT`, so `UPLOAD_URL_PREFIX` is better than `UPLOAD_PATH_PREFIX`.

#### The command

```sh
./answer fix upload PREFIX_SRC PREFIX_DST
```

**Comment by Octobug** [comment]
> Additionally, there is another requirement. We believe it would be helpful to provide an environment variable, `ABSOLUTE_IMAGE_PATH`, which users can set to true. In this way, Answer will default to using absolute paths for storing images. ***This consideration is made for existing users who may copy the content to other places where the images won't display if relative paths are used***.

@LinkinStars I made some attempts on the var part. The consideration you mentioned before (I made it bold & italic) seemd hard to achieve: **The var (`UPLOAD_URL_PREFIX`) is closely associated with [`service_config.upload_path`](https://github.com/apache/incubator-answer/blob/9ebf95cacecf49ff466e47f3ff54c180eb67627c/configs/config.yaml#L35)**.

- If an admin moves the uploaded files **locally**, they need to change `service_config.upload_path` at the same time:
  - And Answer's static router also needs to be changed: <https://github.com/Octobug/incubator-answer/blob/8a8086c87b188e60d9310e234fc3c5353c43f46e/internal/router/static_router.go#L42>
  - Otherwise they need to use other static server (e.g. Nginx) to serve the files. Is this acceptable?
    <img width="1080" alt="image" src="https://github.com/apache/incubator-answer/assets/8007022/bfba26d7-9f5b-48cf-b13a-35ce42a036bf">
  - Now I prefer to tell admins to use a symlink. 🤔️
- If an admin moves the uploaded files to other host that needs to be accessed via an absolute URL, `UPLOAD_URL_PREFIX` doesn't help at all. Because `service_config.upload_path` cannot reach to other host.
  - This var seems to be unnecessary for absolute URLs. Uploading files via plugins also doesn't need it.

**Comment by Octobug** [comment]
> @Octobug 🤔 I got it. So, for now, we can consider only three scenarios:
> 
> 1. Using relative paths: UPLOAD_URL_PREFIX defaults to "/"
> 2. Using absolute paths with the website domain: UPLOAD_URL_PREFIX is "http://answer.cn:8080/"
> 3. Using absolute paths with a third-party service: UPLOAD_URL_PREFIX is "https://s3.east-1.xxx.com/" (Of course, this scene is already handled by the plugin. Just an example.)
> 
> ~This consideration is made for existing users who may copy the content to other places where the images won't display if relative paths are used.~ That would simplify the problem. What do you think?

@LinkinStars 

Right, there are three scenarios for now. But scenario `2.` seems weird 🤔️. What benefits does storing absolute URLs bring (we don't have disagreement on `3.`)? The only benefit I can think of is that admins can use a reverse proxy server (like Nginx) to route the requests of these absolute URLs to where those uploaded files are.

**Comment by Octobug** [comment]
@LinkinStars Aha. Okay, I'll submit a PR a few days later (with the fix command). Thank you for your patience!

**Comment by Octobug** [comment]
@LinkinStars Okay, thank you for your patience! In this case, it seems that your first solution is better. :D That is, make the var a `boolean`. Although the name may need to be changed from `ABSOLUTE_IMAGE_PATH` to `UPLOAD_ABSOLUTE_URL`.

**Comment by Octobug** [comment]
> > * `UPLOAD_USE_ABSOLUTE` (I prefer this one)
> 
> @Octobug 😄 Maybe we can use `UPLOAD_USE_ABSOLUTE_URL`. Thanks for your patience as well. This whole discussion has made the whole issue more clear.

@LinkinStars Aha, okay. I'll submit a PR a few days later together with the `fix` command.

**Comment by Octobug** [comment]
@LinkinStars Here's my version of the `fix` command, I would like to know your opinion. 👀 

`answer fix upload SRC_PREFIX DST_PREFIX [--continue] [--offline|online] [--dry-run] [--limit]`

When the fix command starts, it checks if there is a tmp file: `.fix_upload_interrupted`.

1. If there isn't, it generates one.
2. If there is, it tells the user that the previous fix was interrupted, and the user need to remove the file if they are to start a fresh fix, or use the `--continue` switch to continue.
3. When there isn't any data left that matches the `SRC_PREFIX`, it removes `.fix_upload_interrupted` then exit.

Switches:

- `--continue`: the execution might be interrutped, some of the data is rewritten and the rest of them isn't. It works with the `.fix_upload_interrupted`.
  - All data that matches the `DST_PREFIX` will be ignored.
- `--offline`: it requires the admin to shut down the Answer site, and it works in batch mode.
- `--online`: it allows the fix to happen when a Answer site is up. It fixes a row of data one at a time, and it locks the data preventing it from being load into the web editor, so that the user won't submit the wrong version of links after the fix is done.
  - But this kind of "read lock" seems impossible. I haven't done any research on it. This switch might not be feasible.
- If it is feasible in go command, I would like to implement smaller granularity fix: `fix upload [branding|avatar|quesition]`
- `--dry-run`: only do the test, shows what data will be affected, and shows the "before" and "after", with a `limit` (e.g. first 10 entries).
- `--limit`: fix `limit` rows and exit.

**Comment by Octobug** [comment]
> @Octobug Very thoughtful consideration. I have only one suggestion. I recommend that this command be run offline, but it's not mandatory. It would be advisable to include instructions in the documentation, suggesting users to run it offline. Otherwise, it would require handling various unexpected concurrency situations, such as table locking, which would make the implementation more complex. 🤔️

Okay. I have the feeling that I will give up the `--online` switch. :D

**Comment by Octobug** [comment]
Hi @LinkinStars,

Very sorry for not updating for a long time. I finally have some time now to address this issue.

Last year I wrote some codes for the first part: the installation config - `UPLOAD_USE_ABSOLUTE_URL`: <https://github.com/Octobug/incubator-answer/commits/feat/upload-var-and-fix-cmd/>

But later I thought that maybe it's better to first address this bug itself, that is, to provide the fix command tool. And then in the next step we provide the installation config.

Combining them together makes it too big for a bugfix. What's more, it makes it harder to move forward with resolving this issue.

**Comment by finallyRaunak** [comment]
Hi, slightly off-topic, but I need help configuring Apache Answer. It's only accessible to logged-in, whitelisted users. Here's the setup I'm aiming for:

- **User avatars**: These can be stored in either a public or private S3 bucket.
- **Other media uploads**: These should be stored in a private S3 bucket. When someone visits a post, the media should render using a signed URL with a 30-minute TTL.

I've checked the [documentation](https://github.com/apache/answer-plugins/tree/main/storage-s3), but couldn't find relevant details. Does anyone know how to configure this setup?
